### PR TITLE
sql: add SHOW STATISTICS AS JSON, INJECT STATISTICS

### PR DESCRIPTION
**Pull request by RaduBerinde**
#### sql: add SHOW STATISTICS USING JSON

Release note (sql change): Added SHOW STATISTICS USING JSON, a variant
of SHOW STATISTICS that outputs table statistics as a JSON object; it
can be used to extract statistics from test or customer clusters to
reproduce issues.

#### sql: add ALTER TABLE INJECT STATISTICS

Release note (sql change): Added a new ALTER TABLE command INJECT
STATS which injects table statistics from a JSON object (which can be
obtained via `SHOW HISTOGRAM USING JSON`).

### CODE REVIEW

**Comment by RaduBerinde** [comment]
@knz for grammar changes and inject_stats.go

File: pkg/sql/parser/sql.y [hunk 1]
```diff
@@ -2569,6 +2571,15 @@ zone_value:
     $$.val = tree.NewStrVal($1)
   }
 
+inject_stats_stmt:
+  INJECT STATISTICS FOR TABLE table_name a_expr
```
Comment (petermattis) [inline pkg/sql/parser/sql.y]: I wonder if this should be an `ALTER TABLE` variant. Something like `ALTER TABLE <table> INJECT STATISTICS a_expr`.

Comment (RaduBerinde) [reply pkg/sql/parser/sql.y]: I'm open to that. @knz what do you think?

Also, I'm not sure if I should make the `StatementType` be `Ack` or `DDL` (or what the consequence of choosing one over the other is). Also not entirely sure what privilege to require (currently INSERT; maybe it should be  CREATE like for alter table?).

Note that I plan to add support for this statement in the optimizer test catalog (`exec-ddl` test 
 directive), similar to the support for CREATE TABLE.

**Comment by rytaft** [comment]
:lgtm:

---

Reviewed 4 of 5 files at r1, 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/inject_stats.go, line 37 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9HiEeK-ezByIVZ2tul:-L9HiEeK-ezByIVZ2tum:b-qjns1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ac1ed24a4d7a223a5baedded2672ed7b9eb5eb/pkg/sql/inject_stats.go#L37)):*
> ```Go
> // result of SHOW STATISTICS AS JSON).
> // k$Privileges: INSERT on table.
> func (p *planner) InjectTableStats(
> ```

Can you add a comment about why someone would want to do this (as opposed to CREATE STATISTICS)?

---

*[pkg/sql/inject_stats.go, line 68 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9HjepV4wq9nTcUeljM:-L9HjepV4wq9nTcUeljN:b-2xggzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ac1ed24a4d7a223a5baedded2672ed7b9eb5eb/pkg/sql/inject_stats.go#L68)):*
> ```Go
> 				return nil, fmt.Errorf("statistics cannot be NULL")
> 			}
> 			jsonStr, err := json.Pretty(val.(*tree.DJSON).JSON)
> ```

My json knowledge is pretty minimal, but is there a reason you're calling `json.Pretty` instead of just `json.String()` or something similar?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-:-L9HnWhuDbemKMy9OkXS:bnfp4nl)*

**Comment by RaduBerinde** [comment]
Changed to ALTER TABLE .. INJECT STATISTICS ..

---

Review status: 5 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/inject_stats.go, line 37 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9HiEeK-ezByIVZ2tul:-L9Lan3hCLJ8EMNQNX-I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ac1ed24a4d7a223a5baedded2672ed7b9eb5eb/pkg/sql/inject_stats.go#L37)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a comment about why someone would want to do this (as opposed to CREATE STATISTICS)?
</blockquote></details>

Done.

---

*[pkg/sql/inject_stats.go, line 68 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9HjepV4wq9nTcUeljM:-L9LanzJ9kuCUfWRQOhC:bwvhgmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ac1ed24a4d7a223a5baedded2672ed7b9eb5eb/pkg/sql/inject_stats.go#L68)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

My json knowledge is pretty minimal, but is there a reason you're calling `json.Pretty` instead of just `json.String()` or something similar?
</blockquote></details>

`DJSON.String()` quotes the string; but `.JSON.String()` works.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-:-L9Ln05u7sgQoyHpfQHn:b-i7qe4i)*

**Comment by rytaft** [comment]
*[pkg/sql/inject_stats.go, line 37 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9HiEeK-ezByIVZ2tul:-L9LshB66PAi5yX0fMk0:b75vzfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ac1ed24a4d7a223a5baedded2672ed7b9eb5eb/pkg/sql/inject_stats.go#L37)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Hmm looks like the comment got removed when you copied it over to `alter_table.go`

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488)*

**Comment by RaduBerinde** [comment]
Thanks!

---

Review status: 0 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/alter_table.go, line 702 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9Lp2FN9_jGXbTpPRRu:-L9LsffPAqKAI44Kqhds:b-cy27lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ce4f7ac2005b7889ca8c820de30c4490ba59b26/pkg/sql/alter_table.go#L702)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This might need an additional permission check - are you expecting any user with permission on the table to inject statistics, or only root?
</blockquote></details>

I don't see a good reason to restrict it to root.

---

*[pkg/sql/parser/sql.y, line 2646 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9LoRGa84QwLSW8zPP8:-L9LuPVB6cP3ghUJGrOy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f8df1ad2d64e79e694af9e11512c3e0e14cf6b9/pkg/sql/parser/sql.y#L2646)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The correct keyword for options in SQL is `USING`, not `AS`.

(`AS` serves exclusively to establish equivalence between a name and some other thing, ie. establish aliases, define CTEs etc)
</blockquote></details>

Done.

---

*[pkg/sql/inject_stats.go, line 37 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9HiEeK-ezByIVZ2tul:-L9LvZGL4s0v28TVYqzV:b8gj3mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ac1ed24a4d7a223a5baedded2672ed7b9eb5eb/pkg/sql/inject_stats.go#L37)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Hmm looks like the comment got removed when you copied it over to `alter_table.go`
</blockquote></details>

Good catch, fixed.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-:-L9LuUY48A_t-3JIKgJf:b-ucfknh)*

**Comment by knz** [comment]
:lgtm: but you need to update both commit messages to add a release note. I'd suggest converting the bulk of the commit messages to release notes altogether (Put the `Release note (sql change):` prefix in front of them)

---

Reviewed 1 of 15 files at r4, 14 of 14 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/alter_table.go, line 702 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-L9Lp2FN9_jGXbTpPRRu:-L9Lw_7C1Iu8b0cLokWB:bwzwlhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ce4f7ac2005b7889ca8c820de30c4490ba59b26/pkg/sql/alter_table.go#L702)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't see a good reason to restrict it to root.
</blockquote></details>

All right.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24488#-:-L9LwoKh3-wp_sAfFSY1:b910q68)*

**Comment by RaduBerinde** [comment]
Thanks, done.

**Pull request merged**